compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -4041,6 +4095,16 @@ public void visit(BLangConstant constant, AnalyzerData data) {
         constantAnalyzer.visit(constant);
     }
 
+    private boolean isNodeKindAllowedForConstants(BLangExpression expression, BLangConstant constant) {
+        return ((expression.getKind() == NodeKind.UNARY_EXPR &&
+                ((BLangUnaryExpr) constant.expr).expr.getKind() != NodeKind.NUMERIC_LITERAL) &&
+                (!OperatorKind.SUB.equals(((BLangUnaryExpr) constant.expr).operator) ||
+                        !OperatorKind.ADD.equals(((BLangUnaryExpr) constant.expr).operator)) ||
+                !(expression.getKind() == LITERAL || expression.getKind() == NUMERIC_LITERAL) &&
+                        expression.getKind() != NodeKind.UNARY_EXPR) &&
+                constant.typeNode == null;
```

Comment:
Can improve readability by breaking into multiple if conditions and returning early.